File: app/src/main/java/com/example/expense_management/RecyclerViewAdapters/mainRecycleAdapter.java
package com.example.expense_management.RecyclerViewAdapters;

import android.text.format.DateFormat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.view.animation.ScaleAnimation;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.expense_management.Database.DataItems;
import com.example.expense_management.R;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import static java.lang.String.*;

public class mainRecycleAdapter extends RecyclerView.Adapter<mainRecycleAdapter.ViewHolder> {

    List<DataItems> dataItemsList;
    private onItemClickListener listener;
    private onItemLongClickListener longListener;

    public mainRecycleAdapter(){

    }
    public void setDataItemsList(List<DataItems> dataItemsList){
        this.dataItemsList=new ArrayList<>();
        if(dataItemsList.isEmpty()){
            this.dataItemsList=new ArrayList<>();   //setting dataItemsList
        }
        this.dataItemsList.clear();
        this.dataItemsList.addAll(dataItemsList);
        notifyDataSetChanged();
    }
    public DataItems getItemList(int position){
        return dataItemsList.get(position);
    }
    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view= LayoutInflater.from(parent.getContext()).inflate(R.layout.recycler_view,parent,false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
            TextView date=holder.date;
            String stringDate= DateFormat.format("dd/MM/yy",new Date(dataItemsList.get(position).getDate())).toString();
            date.setText(stringDate);
            TextView gainMoney=holder.gain;
            gainMoney.setText(""+dataItemsList.get(position).getGrossMoneyGot());
            TextView paidMoney=holder.paid;
            paidMoney.setText(format("%d", dataItemsList.get(position).getGrossMoneyExpense()));
            setFadeAnimation(holder.itemView);
    }

    @Override
    public int getItemCount() {
        if (dataItemsList!=null)
            return dataItemsList.size();
        return 0;
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
       public TextView date,gain,paid;
        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            date=itemView.findViewById(R.id.date_of_data);
            gain=itemView.findViewById(R.id.gain_data);
            paid=itemView.findViewById(R.id.paid_data);
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    int position=ViewHolder.this.getAdapterPosition();
                    if (listener!=null&&position!=RecyclerView.NO_POSITION)
                        listener.onItemClicked(dataItemsList.get(position));
                }
            });
            itemView.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    int position=ViewHolder.this.getAdapterPosition();
                    if (longListener!=null&&position!=RecyclerView.NO_POSITION)
                        longListener.onItemLongClicked(dataItemsList.get(position));
                    return true;
                }
            });
        }
    }
    public void setOnItemLongClickListener(onItemLongClickListener longListener){
        this.longListener=longListener;
    }
    public interface onItemLongClickListener {
        void onItemLongClicked(DataItems data);
    }
    public void setOnItemClickListener(onItemClickListener listener){
        this.listener=listener;
    }
    public interface onItemClickListener {
        void onItemClicked(DataItems data);
    }
    public static void setFadeAnimation(View view){
        ScaleAnimation alphaAnimation=new ScaleAnimation(0.0f,
                1.0f, 0.0f, 1.0f, Animation.RELATIVE_TO_SELF,
                0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
        alphaAnimation.setDuration(1000);
        view.startAnimation(alphaAnimation);
    }
}

File: app/src/main/java/com/example/expense_management/recyclerViewAdapters/MainRecycleAdapter.kt
package com.example.expense_management.recyclerViewAdapters

import android.text.format.DateFormat
import androidx.recyclerview.widget.RecyclerView
import com.example.expense_management.database.DataItems
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import com.example.expense_management.R
import android.widget.TextView
import android.view.animation.ScaleAnimation
import android.view.animation.Animation
import java.util.*

class MainRecycleAdapter : RecyclerView.Adapter<MainRecycleAdapter.ViewHolder>() {
    var dataItemsLt: MutableList<DataItems>? = null
    private var listener: OnItemClickListener? = null
    private var longListener: OnItemLongClickListener? = null
    fun setDataItemsList(dataItemsList: List<DataItems>) {
        this.dataItemsLt = ArrayList()
        if (dataItemsList.isEmpty()) {
            this.dataItemsLt = ArrayList() //setting dataItemsList
        }
        (this.dataItemsLt as ArrayList<DataItems>).clear()
        (this.dataItemsLt as ArrayList<DataItems>).addAll(dataItemsList)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view =
            LayoutInflater.from(parent.context).inflate(R.layout.recycler_view, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val date = holder.date
        val stringDate = DateFormat.format(
            "dd/MM/yy", Date(
                dataItemsLt!![position].date
            )
        ).toString()
        date.text = stringDate
        val gainMoney = holder.gain
        gainMoney.text = "" + dataItemsLt!![position].grossMoneyGot
        val paidMoney = holder.paid
        paidMoney.text = String.format("%d", dataItemsLt!![position].grossMoneyExpense)
        setFadeAnimation(holder.itemView)
    }

    override fun getItemCount(): Int {
        return if (dataItemsLt != null) dataItemsLt!!.size else 0
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var date: TextView = itemView.findViewById(R.id.date_of_data)
        var gain: TextView = itemView.findViewById(R.id.gain_data)
        var paid: TextView = itemView.findViewById(R.id.paid_data)

        init {
            itemView.setOnClickListener {
                val position = this@ViewHolder.adapterPosition
                if (listener != null && position != RecyclerView.NO_POSITION) listener!!.onItemClicked(
                    dataItemsLt!![position]
                )
            }
            itemView.setOnLongClickListener {
                val position = this@ViewHolder.adapterPosition
                if (longListener != null && position != RecyclerView.NO_POSITION) longListener!!.onItemLongClicked(
                    dataItemsLt!![position]
                )
                true
            }
        }
    }

    fun setOnItemLongClickListener(longListener: OnItemLongClickListener) {
        this.longListener = longListener
    }

    interface OnItemLongClickListener {
        fun onItemLongClicked(data: DataItems)
    }

    fun setOnItemClickListener(listener: OnItemClickListener) {
        this.listener=listener
    }

    interface OnItemClickListener {
        fun onItemClicked(data: DataItems)
    }

    companion object {
        fun setFadeAnimation(view: View) {
            val alphaAnimation = ScaleAnimation(
                0.0f,
                1.0f, 0.0f, 1.0f, Animation.RELATIVE_TO_SELF,
                0.5f, Animation.RELATIVE_TO_SELF, 0.5f
            )
            alphaAnimation.duration = 1000
            view.startAnimation(alphaAnimation)
        }
    }
}